Adds the how it works page for spfx project upgrade (#5216)
pnp/cli-microsoft365: docs/docs/concepts/how-it-works/spfx-project-upgrade.mdx

> # How it works: spfx project upgrade

Reviewer: Let's add a front-matter title so it appears better in the sidebar navigation. That way, under the section `How it works`, it will show as `spfx project upgrade` instead of `How it works: spfx project upgrade`

```suggestion
---
title: spfx project upgrade
---

# How it works: spfx project upgrade
```

## Concepts

Following concepts are fundamental for understanding how the `spfx project upgrade` command works.

### Supported SPFx version

Upgrading SPFx projects from one version to another often requires more changes than just upgrading core SPFx packages. When a new SPFx version is released, we analyze the changes between the latest and the previous version of SPFx and describe them using a **version delta**. To let our users know that we support upgrading SPFx projects to that version, we add it to the array of supported versions in the `src/m365/spfx/commands/project/project-upgrade.ts` file.

Supported versions are ordered in chronological order in which they were released. This is important, because it determines the order in which rules should be applied when upgrading SPFx projects across multiple versions.

### Version delta

The `spfx project upgrade` command uses version deltas to describe changes between SPFx versions. Each delta describes changes between two versions: the version to upgrade to and the previous version, for example: the `upgrade-1.17.4.ts` delta describes only what's changed between SPFx versions 1.17.3 and 1.17.4.

> Breaking down deltas to just two SPFx versions allows users to upgrade their projects in any supported SPFx range, eg. 1.4.1 to 1.16.0 or 1.14.0 to the latest SPFx version supported by the `spfx project upgrade` command.

Reviewer: ```suggestion
Breaking down deltas to just two SPFx versions allows users to upgrade their projects in any supported SPFx range, e.g. 1.4.1 to 1.16.0 or 1.14.0 to the latest SPFx version supported by the `spfx project upgrade` command.
```

To allow rules to be applied to different versions of SPFx, they can accept parameters, such as a dependency version number, property value, etc.

Rules are stored in `src/m365/spfx/commands/project/project-upgrade/rules`. Each rule has a unique identifier which allows us to quickly identify related rules. A rule identifier is built as follows: `FNXXXYYY`, where:

- `FN` is the fixed prefix for _finding_
- `XXX` is the three digit number of the type of finding, eg. dependency rules are 001, dev dependencies 002, etc.
- `YYY` is the three digit number of the rule in its group

Rules' filenames use the following naming convention: `FNXXXYYY_ZZZ_name`, where:

- `FNXXXYYY` is the rule's identifier
- `ZZZ` is the rule's group abbreviation for easy lookup
- `name` is the rule's name for easy lookup

Across the different versions of SPFx, we've built rules that should cover for the typical changes in projects between SPFx versions. That said, in case there's a need for a new rule, follow the design pattern of other rules and reach out to CLI maintainers to verify your design.

### Superseded rules

> A rule can supersede one or more other rules. You define superseded rules, by overriding the `supersedes` getter in the rule and return the array of rule IDs that the rule supersedes, eg.

Reviewer: ```suggestion
A rule can supersede one or more other rules. You define superseded rules, by overriding the `supersedes` getter in the rule and return the array of rule IDs that the rule supersedes, e.g.
```

To speed up the upgrade process, the `spfx project upgrade` command starts the upgrade by collecting the relevant project files and building an object model out of them. That way, files need to be parsed only once and rules can be applied on in-memory objects that represent the project. This makes the upgrade process fast, even when it spans multiple versions with many different changes.

Project model is stored in `src/m365/spfx/commands/project/project-model`.

### Exit codes

Shell executables (CLIs, scripts, etc.) communicate to the parent process whether their execution succeeded using exit codes. An exit code of 0 means success and a non-0 exit code means failure. The `spfx project upgrade` command uses exit codes to communicate different error states to the parent process using it. These error codes have been originally designed for use with the PnP Yeoman Generator.

The exit codes of the `spfx project upgrade` command are defined in the `src/m365/spfx/commands/project/project-upgrade.ts` file.

## How it works

This section explains the process that the `spfx project upgrade` command follows to produce steps to upgrade the SPFx project to the specified SPFx version.

> ![Diagram illustrating the process of the spfx project upgrade command](../../images/how-it-works/spfx-project-upgrade-process.png)

Reviewer: Would it be possible to have a different image here in dark mode? Currently, this hurts my eyes a bit 😄 

Another suggestion would be to scale this image down a bit as it's quite a big image to view on the page. Could we scale it down so it can be fully visible on the screen without the need or scrolling?

Author: Is there a way for us to switch images in the content between the two modes?

Reviewer: Just had a quick look and there appears to be a component named `ThemedImage` we can use. Ref: https://docusaurus.io/docs/markdown-features/assets#themed-images

Author: Do you know by any chance if Docusaurus supports mermaid diagrams? It would make it easier for us to maintain it long-term as it's just code and perhaps would give us the ability to render light- and dark-mode without keeping two different images.

Here's the mermaid flowchart that we could use:

```mermaid
flowchart TD
    start(start) --> isSpfx{Executed on an SPFx project?}
    isSpfx -- no, exit: 1 --> stop(end)
    isSpfx -- yes --> isToVersionSupported{Is the version to upgrade to supported?}
    isToVersionSupported -- no, exit: 2 --> stop
    isToVersionSupported -- yes --> detectVersion(Detect SPFx version)
    detectVersion --> isVersionDetected{Version detected?}
    isVersionDetected -- no, exit: 3 --> stop
    isVersionDetected -- yes --> isFromVersionSupported{Is the from version supported?}
    isFromVersionSupported -- no, exit: 4 --> stop
    isFromVersionSupported -- yes --> isToVersionLarger{Is the to version larger than from version?}
    isToVersionLarger -- no, exit: 5 --> stop
    isToVersionLarger -- yes --> areToAndFromTheSame{Are the from and to versions the same?}
    areToAndFromTheSame -- yes --> stop
    areToAndFromTheSame -- no --> collect(Collect project)
    collect --> deltas(Calculate deltas)
    deltas --> applyRules(Apply upgrade rules)
    applyRules --> dedupe(Deduplicate findings)
    dedupe --> removeSuperseded(Remove superseded findings)
    removeSuperseded --> flatten(Flatten findings)
    flatten --> replaceTokens(Replace tokens in resolutions)
    replaceTokens --> buildReport(Build the report)
    buildReport --> stop
```

Reviewer: That's a great idea. Just had a quick glance and it appears that we can enable it https://docusaurus.io/docs/markdown-features/diagrams

Author: Should we add it to this PR or would you rather like to keep it separate? If we add this to this PR, you could submit a commit with the necessary changes to this PR

Author: I've updated the PR with support for mermaid

### Detecting project version

The `spfx project upgrade` command starts by checking if it's been executed in a folder belonging to an SPFx project.

> Once the command located the SPFx project's root folder, it will try to determine the project's version. It does this by using the following steps:

Reviewer: ```suggestion
Once the command locates the SPFx project's root folder, it will try to determine the project's version. It does this by using the following steps:
```

1. Check for the presence of the `.yo-rc.json` file. If found, get the value of the `@microsoft/generator-sharepoint.version` property. When present, check the value of the `@microsoft/generator-sharepoint.environment` property to adjust the version number for SPFx versions targeting on-prem SharePoint deployments.
1. If `.yo-rc.json` was not found, check for the presence of the `package.json` file. If found, read the version of the `@microsoft/sp-core-library` dependency.

> After it determined the project version, the command will verify that it supports the specified version, and the specified version to upgrade the project to is newer than the detected project version.

Reviewer: ```suggestion
After it determines the project version, the command will verify that it supports the specified version, and the specified version to upgrade the project to is newer than the detected project version.
```

### Collecting the project to upgrade

Next, the command will parse the project to upgrade and build an object model off of it. It will use this model to detect which changes need to be applied to the project to make it match a newly created SPFx version of the specified version.

### Calculating upgrade deltas

As the next step, the command determines the upgrade deltas: the list of upgrades and rules to upgrade the project from the version the command detected to the version specified by the user (or the latest supported version if the user hasn't specified any version).

For example, a user wants to upgrade their SPFx project from version 1.17.0 to 1.17.4. The `spfx project upgrade` command will load deltas `1.17.1`, `1.17.2`, `1.17.3` and `1.17.4`, and stack their rules in the order they're defined in the delta files, so that rules defined in `1.17.4` are executed as last. This will make them take precedence in the **deduplication stage**.

### Applying upgrade rules

> After the command built the list of rules to upgrade the project to the specified SPFx version, it executes them in the order they're defined in deltas.

Reviewer: ```suggestion
After the command builds the list of rules to upgrade the project to the specified SPFx version, it executes them in the order they're defined in deltas.
```
